### ISSUE TITLE
Some operation on Diagram seems extremely slow

### ISSUE THREAD

**Issue by wkoiking** [cchalmers/diagrams-dev#2]
Following code takes more than 5 min to generate the diagram:

~~~
renderTest :: IO ()
renderTest = saveDiagram Cairo "XXX.png" (mkWidth 1000) $ mconcat
    [ rect 20 20
    , center (withEnvelope alignedDiagramL (mconcat [leftSignalDrawing, alignedDiagramL]))
    ]
 where alignedDiagramL = alignL $ withEnvelope alignedDiagramR
                                $ mconcat [rightSignalDrawing, alignedDiagramR]
       alignedDiagramR = alignR $ withEnvelope alignedDiagramC
                                $ mconcat [stationDrawing, alignedDiagramC]
       alignedDiagramC = centerX trackDrawing
       -- basic drawing
       trackDrawing       = rect 10 1 # fc yellow
       leftSignalDrawing  = rotate (90 @@ deg) $ triangle 1 # fc gray
       rightSignalDrawing = rotate (- 90 @@ deg) $ triangle 1 # fc gray
       stationDrawing     = center $ vcat $ replicate n
                                   $ hcat $ replicate n $ circle 0.2 # fc blue
        where n = 10

~~~

While rendering of each "basic drawing" finishes instantly, i.e., following code will return instantly:
~~~
renderTest :: IO ()
renderTest = saveDiagram Cairo "XXX.png" (mkWidth 1000) $ center $ hsep 1
    [ trackDrawing
    , leftSignalDrawing
    , rightSignalDrawing
    , stationDrawing
    ]
 where -- basic drawing
       trackDrawing       = rect 10 1 # fc yellow
       leftSignalDrawing  = rotate (90 @@ deg) $ triangle 1 # fc gray
       rightSignalDrawing = rotate (- 90 @@ deg) $ triangle 1 # fc gray
       stationDrawing     = center $ vcat $ replicate n
                                   $ hcat $ replicate n $ circle 0.2 # fc blue
        where n = 10
~~~

In the diagrams-1.4, the former code also returns instantly. So I assume this to be the problem of diagrams-2.0. 

The former code generates following image (when n = 5):
![XXX](https://user-images.githubusercontent.com/3189740/55451310-5d379c80-560d-11e9-87a1-482d1c78424d.png)

Environment:
Compiler: GHC 8.4.4
OS: Windows 10

**Comment by cchalmers** [comment]
Hmm, this is a weird one. Just by adding `:: Diagram V2` to `alignedDiagramC` gives a 10x speedup. Even though that's the only type it could be (I don't have `NoMonomorphismRestriction` on), so I'm guessing it's a specialisation not firing. @byorgey this is one of the reasons I want to get rid of the `n` parameter in diagrams.

I managed to trim down the rest of the slowdown to a `withEnvelope alignedDiagramL alsignedDiagramL` being 30x slower than just `alignedDiagramL`. Which is super weird, I'm still trying to see what's causing that.

**Comment by cchalmers** [comment]
Hmm, I think I'm just being silly when I get the envelope of a diagram. I visit all the nodes even if I've replaced the envelope of a part of it. I need to think about the best way to solve this. I might just write a specialsed `getEnvelope` for diagrams instead of using `fldU`.

**Comment by cchalmers** [comment]
Fixed with cchalmers/diagrams@4f7ff3e3c7f862dd9404eb0b62ce68fd4d536b41. This runs in 0.054 seconds for me. Thanks for the report!

**Comment by wkoiking** [comment]
I have also checked and confirmed that my application (based on diagrams-2.0) starts in a second. Thanks!
